app/merginapi.cpp

```diff
@@ -5,29 +5,101 @@
 #include <QJsonObject>
 #include <QJsonArray>
 #include <QDate>
+#include <QByteArray>
+#include <QSet>
 
+#ifdef ANDROID
+#include <QtAndroid>
+#include <QAndroidJniObject>
+#endif
 
-MerginApi::MerginApi(const QString &root, QByteArray token, QObject *parent)
+MerginApi::MerginApi(const QString &root, const QString& dataDir, QByteArray token, QObject *parent)
   : QObject (parent)
   , mApiRoot(root)
+  , mDataDir(dataDir + "/downloads/")
   , mToken(token)
 {
 }
 
 void MerginApi::listProjects()
 {
+    if (mToken.isEmpty()) {
+        emit networkErrorOccurred( "Auth token is invalid", "Mergin API error: listProjects" );
+        return;
+    }
+
     mMerginProjects.clear();
     QNetworkRequest request;
     // projects filtered by tag "input_use"
     QUrl url(mApiRoot + "/v1/project?tags=input_use");
-
     request.setUrl(url);
     request.setRawHeader("Authorization", QByteArray("Basic " + mToken));
 
     QNetworkReply *reply = mManager.get(request);
     connect(reply, &QNetworkReply::finished, this, &MerginApi::listProjectsReplyFinished);
 }
 
+void MerginApi::downloadProject(QString projectName)
+{
+    if (mToken.isEmpty()) {
+        emit networkErrorOccurred( "Auth token is invalid", "Mergin API error: downloadProject" );
+    }
+
+    QNetworkRequest request;
+    QUrl url(mApiRoot + "/v1/project/download/" + projectName);
+
+    if (mPendingRequests.contains(url)) {
+        QString errorMsg = QStringLiteral("Download request for %1 is already pending.").arg(projectName);
+        qDebug() << errorMsg;
+        emit networkErrorOccurred( errorMsg, "Mergin API error: downloadProject" );
+        return;
+    }
+
+    request.setUrl(url);
+    request.setRawHeader("Authorization", QByteArray("Basic " + mToken));
+
+    QNetworkReply *reply = mManager.get(request);
+    mPendingRequests.insert(url, projectName);
+    connect(reply, &QNetworkReply::finished, this, &MerginApi::downloadProjectReplyFinished);
+}
+
+void MerginApi::getInfoAndFetch(QString projectName)
+{
+
+    if (mToken.isEmpty()) {
+        emit networkErrorOccurred( "Auth token is invalid", "Mergin API error: projectInfo" );
+    }
+
+    QNetworkRequest request;
+    QUrl url(mApiRoot + "/v1/project/" + projectName);
+
+    request.setUrl(url);
+    request.setRawHeader("Authorization", QByteArray("Basic " + mToken));
+
+    QNetworkReply *reply = mManager.get(request);
+    connect(reply, &QNetworkReply::finished, this, &MerginApi::projectInfoReplyFinished);
+
+}
+
+void MerginApi::fetchProject(QString projectName, QByteArray json)
```

Comment:
can we rename to something like downloadProjectFiles() ? I was having trouble understanding the difference between downloadProject() and fetchProject() - from the name they seem to do the same thing.